src/Functions/tokens.cpp

```diff
@@ -76,13 +80,47 @@ class FunctionTokens : public IFunction
                                                         : arguments[arg_tokenizer].column->getDataAt(0).toView();
 
         if (tokenizer_arg == SplitTokenExtractor::getExternalName())
+        {
             token_extractor = std::make_unique<SplitTokenExtractor>();
+        }
+        else if (tokenizer_arg == StringTokenExtractor::getExternalName())
+        {
+            std::vector<String> separators;
+            if (arguments.size() < 3)
+                separators = {" "};
+            else
+            {
+                const ColumnArray * col_separators_non_const = checkAndGetColumn<ColumnArray>(arguments[arg_separators].column.get());
+                const ColumnConst * col_separators_const = checkAndGetColumnConst<ColumnArray>(arguments[arg_separators].column.get());
+
+                if (col_separators_const)
+                {
+                    const Array & separators_array = col_separators_const->getValue<Array>();
+                    for (const auto & separator : separators_array)
+                        separators.emplace_back(separator.safeGet<String>());
+                }
+                else if (col_separators_non_const)
+                {
+                    Field separator_field;
+                    col_separators_non_const->get(0, separator_field);
+                    Array & separator_array = separator_field.safeGet<Array>();
+                    for (const auto & separator : separator_array)
+                        separators.emplace_back(separator.safeGet<String>());
+                }
+                else
+                {
+                    throw Exception(ErrorCodes::ILLEGAL_COLUMN, "3rd argument of function {} should be Array(String), got: {}", name, arguments[arg_separators].column->getFamilyName());
```

Comment:
Shouldn't this be a logical error? AFAIU `getReturnTypeImpl` should make sure the argument types are correct.

Reply:
You are right (kind of). The thing is that `validateFunctionArguments` in `getReturnTypeImpl` can only check that the outer type is expected (`Array`) but not the inner type (`Array(String)`). This is a known limitation.

It could be fixed but that wouldn't make the code more beautiful as we'd still need to check in two places. So I'd vote for keeping everything as is.